common/remote_object.go

```diff
@@ -0,0 +1,128 @@
+/*
+ *
+ * xk6-browser - a browser automation extension for k6
+ * Copyright (C) 2021 Load Impact
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Affero General Public License as
+ * published by the Free Software Foundation, either version 3 of the
+ * License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package common
+
+import (
+	"context"
+	"encoding/json"
+	"math"
+	"strconv"
+	"strings"
+
+	cdpruntime "github.com/chromedp/cdproto/runtime"
+	"github.com/dop251/goja"
+	"github.com/sirupsen/logrus"
+	k6common "go.k6.io/k6/js/common"
+)
+
+type objectOverflowError struct{}
+
+// Error returns the description of the overflow error.
+func (*objectOverflowError) Error() string {
+	return "object is too large and will be parsed partially"
+}
+
+func parseRemoteObjectPreview(op *cdpruntime.ObjectPreview, logger *logrus.Entry) (map[string]interface{}, error) {
+	obj := make(map[string]interface{})
+	var err error
+	if op.Overflow {
+		err = &objectOverflowError{}
+	}
+
+	for _, p := range op.Properties {
+		val, err := parseRemoteObjectValue(p.Type, p.Value, p.ValuePreview, logger)
+		if err != nil {
+			logger.WithError(err).Errorf("failed parsing object property %q", p.Name)
+			continue
+		}
+		obj[p.Name] = val
+	}
+
+	return obj, err
+}
```

Comment:
I agree that asserting on log messages in tests is brittle, but as to which component can log is completely arbitrary. :man_shrugging:

But sure, I'll remove it from here as well.

Reply:
One more idea if you fancy this one more:

```golang
func parseRemoteObjectPreview(op *cdpruntime.ObjectPreview) (map[string]interface{}, []string, error) {
	var (
		warnings []string
		obj = make(map[string]interface{})
	)
	if op.Overflow {
		warnings = append(warnings, "object is too large and will be parsed partially")
	}
	for _, p := range op.Properties {
		val, err := parseRemoteObjectValue(p.Type, p.Value, p.ValuePreview)
		if err != nil {
			warnings = append(warnings, fmt.Sprintf("failed parsing object property %q", p.Name))
			continue
		}
		obj[p.Name] = val
	}

	return obj, warnings, err
}
```

```golang
	for _, robj := range event.Args {
		i, warns, err := parseRemoteObject(robj)
		if err != nil {
			// If this throws it's a bug :)
			k6Throw(fs.ctx, "unable to parse remote object value: %w", err)
		}
		for _, w := range warns {
			l.Warn(w)
		}
		parsedObjects = append(parsedObjects, i)
	}
```

Reply:
@inancgumus Take a look at 7a5de2f. I decided to use `multierror` after all, the API is nice and I don't want to reinvent the wheel. I wouldn't want to add a separate return argument for warnings, especially since failing to parse an object should be handled (logged) as an error.

I'm having an issue with the error assertion in the test, maybe you can help me out, see my other comment.

Reply:
Yeah, good call 😄 👍  I added my suggestions that fixes the problem.